### PR TITLE
Add SortFlags option

### PR DESCRIPTION
**Pull request by bogem**
Updates https://github.com/spf13/cobra/issues/316

### CODE REVIEW

File: flag.go [hunk 1]
```diff
@@ -194,10 +199,12 @@ func sortFlags(flags map[NormalizedName]*Flag) []*Flag {
 // "--getUrl" which may also be translated to "geturl" and everything will work.
 func (f *FlagSet) SetNormalizeFunc(n func(f *FlagSet, name string) NormalizedName) {
 	f.normalizeNameFunc = n
+	f.order = f.order[:0]
```
Comment (eparis) [inline flag.go]: Is this some shorthand for `f.order = []NormalizedName{}` ?

File: flag.go [hunk 2]
```diff
@@ -232,8 +239,14 @@ func (f *FlagSet) SetOutput(output io.Writer) {
 // VisitAll visits the flags in lexicographical order, calling fn for each.
 // It visits all flags, even those not set.
 func (f *FlagSet) VisitAll(fn func(*Flag)) {
```
Comment (eparis) [inline flag.go]: Why only `VistAll` and not `Visit` ?  Should we really change the behavior of `sortFlags()`

File: flag.go [hunk 3]
```diff
@@ -134,10 +134,15 @@ type FlagSet struct {
...
 
+	// SortFlags is used to indicate, if user wants to have sorted flags in
+	// help/usage messages.
+	SortFlags bool
+
 	name              string
 	parsed            bool
 	actual            map[NormalizedName]*Flag
 	formal            map[NormalizedName]*Flag
+	order             []NormalizedName
```
Comment (eparis) [inline flag.go]: I'm not sure I like how this works. I'm wondering if it should just be a []*Flag and be used other places...   But I'm unsure...

Comment (bogem) [inline flag.go]: Yes, it would be to have array of flags than array of names. I didn't think about it

File: flag.go (hunk 1 above)
Comment (bogem) [inline flag.go]: When you do `f.order = []NormalizedName{}` you create one more array and make one more allocation. But if you do `f.order = f.order[:0]` you reuse the current array but just delete all elements in it.

**Comment by bogem** [comment]
> I'm not sure I like how this works. I'm wondering if it should just be a []*Flag and be used other places... But I'm unsure...
Yes, it would be better to have an slice of flags rather than slice of names. I didn't think about it

> Is this some shorthand for f.order = []NormalizedName{} ?
When you do f.order = []NormalizedName{} you create one more array and make one more allocation. But if you do f.order = f.order[:0] you reuse the current array but just delete all elements in slice.
 
> Why only VistAll and not Visit ? Should we really change the behavior of sortFlags()
I just forgot about Visit.

**Comment by bogem** [comment]
Everything is fixed according to your correctives

**Comment by eparis** [comment]
I know I made those comments, and from a code PoV it looks fine. I'm wondering if we should instead take a sort func? What if someone wants something other than 'alphabetical' and 'order it was created' ? We haven't had many people complain about sorting, but codifying those 2 and only those 2 as pflag API rubs me the wrong way.

Do you think a function would be a better API? Or just needless flexibility?

**Comment by eparis** [comment]
(also maybe a function wouldn't be possible since too much of the requisite information is in private fields)

**Comment by bogem** [comment]
It's a great idea but IMO, it will be redundant flexibility and also a complexity to library.
And actually how could be sorted flags, except alphabetically and order it was created?

**Comment by eparis** [comment]
@bogem I can't think of a good way, but I don't want to be too limiting if it exists. For big projects with flags coming from a large number of places (but explicit and from importing go flags as persistent flags) controlling order can be really hard as things get refactored.

I guess we can go with this. If anyone ask for a function we can still do it.

**Pull request merged**